### PR TITLE
Setup project

### PR DESCRIPTION
**Pull request by ramobis**
I added all the related source code and a compact documentation about conceptional and technical project details.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,57 +1,348 @@
...
+# green-path-optimizer [@OST](https://www.ost.ch/en/)
 
-An application that suggests changes to the forwarding paths of a network to improve energy efficiency developed at [RIPE NCC Green Tech Hackathon](https://labs.ripe.net/author/becha/announcing-the-green-tech-hackathon/).
+Is an application framework for simulating a proof-of-concept environment that demonstrates an approach to gaining transparency on the environmental costs of network paths and optimising the flow of data within that network to reduce the carbon footprint caused by its operation.
 
-## Overall Project Achievements
+**Currently** it contains the following components: 
 
-### Before Hackathon
+- **Network** based on BMv2 software switches capable of:
```
Comment (severindellsperger) [inline README.md:17]: Proposal: Add link of BMv2 software switches

File: README.md [hunk 2]
```diff
@@ -1,57 +1,348 @@
...
+  - Collecting inband network telemetry (INT) data regarding energy efficiency of network paths using the IOAM protocol
+  - Exporting the INT data using IPFIX
```
Comment (severindellsperger) [inline README.md]: Add IPFIX RFC

File: README.md [hunk 3]
```diff
@@ -1,57 +1,348 @@
...
-  - Interpretation of improvements
-  - Validation of the suggested improvements also regarding throughput / bottlenecks etc.
+- **Optimizer** to be developed at [RIPE NCC Green Tech Hackathon](https://labs.ripe.net/author/becha/announcing-the-green-tech-hackathon/) which is capable of:
+  - Analyzing the given time series data stored in InfluxDB
+  - Proposing configuration changes to route traffic over the most efficient paths between every possible ingress and egress router
+  - Trigger configuration update to actually implement the proposed optimization in the network.
+- **Validator** which is to be developed after the hackathon and validates the optimiser's configuration changes to e.g. prevent data from being sent through bottlenecks or certain paths from being overprovisioned.
 
-### After Hackathon
+## Contents
```
Comment (severindellsperger) [inline README.md]: I would probably start with the content section. The part above could be within an Introduction section. So, the reader can jump to the appropriate chapter directly.

File: README.md [hunk 4]
```diff
@@ -1,57 +1,348 @@
...
+  - Parsing of IOAM raw export IPFIX messages ([Telegraf](https://www.influxdata.com/time-series-platform/telegraf/))
+  - Persistent storage of efficiency data ([InfluxDB](https://www.influxdata.com/))
+  - Visualization of efficiency data ([Grafana](https://grafana.com/))
 
-- Dynamic Grafana Dashboards to allow simulations on arbitrary topologies
-- Skeleton of green-path-optimizer application
-  - Retrieve data from InfluxDB
-- Implementation of flooding mechanism for probing packets
-  - Loops? / Spanning Tree
+**In near futre** it will also contain the following components:
```
Comment (severindellsperger) [inline README.md]:
```suggestion
**In near future** it will also contain the following components:
```

File: README.md [hunk 5]
```diff
@@ -1,57 +1,348 @@
...
-After several months of research, we published a paper titled "Towards Sustainable Networking: Unveiling Energy Efficiency Through Hop and Path Efficiency Indicators in Computer Networks", which was presented at IEEE Netsoft 2024 and is now available on [IEEE Xplore](https://ieeexplore.ieee.org/document/10588907).
+After several months of research, we published a paper titled *"Towards Sustainable Networking: Unveiling Energy Efficiency Through Hop and Path Efficiency Indicators in Computer Networks"*, which was presented at IEEE Netsoft 2024 and is now available on [IEEE Xplore](https://ieeexplore.ieee.org/document/10588907).
+
+Our proof-of-concept (PoC) implementation already includes several key capabilities:
+- The collection of network telemetry data embedded in packet metadata using the [IOAM Aggregation Trace Option](https://datatracker.ietf.org/doc/html/draft-cxx-ippm-ioamaggr-02)
+- Standardized export of the collected network telemetry data to an [IPFIX](https://www.rfc-editor.org/rfc/rfc7011) collector
+- Visualization of the collected data on descriptive Grafana dashboards
```
Comment (severindellsperger) [inline README.md:51]: I would link a picture so that the user can get an idea of what the visualization dashboard looks like.

File: README.md [hunk 6]
```diff
@@ -1,57 +1,348 @@
...
+- **Hop Efficiency Indicator (HEI):** Is an arbitrary number indicating the efficiency of a hop. There can be several HEI values at the same time, which cover different aspects of a hop's energy efficiency.
+- **Link Efficiency Indicator (LEI):** Is a dedicated value to indicate the efficiency of an interface.
+For example a 10GBit/s copper interface via a twisted pair cable could be indicated to be less expensive in the means of energy efficiency compared to an interface connected to a 10GBit/s long-haul fiber connection.
+- **Hop Traversal Cost (HTC):** Is the result of accumulating the LEI of the ingress link, the HEI and the LEI of the egress link.
+- **Path Efficiency Indicator (PEI):** Is the accumulation of HTC values in case the SUM aggregator is used. It indicates the efficiency of the path the packet traversed.
+- **Flow Efficiency Indicator (FEI):** Is depending on the aggregator used the average PEI (SUM aggregator), the minimum HEI (MIN aggregator) or maximum HEI (MAX aggregator) considering the network telemetry data of all packets corresponding to the specific flow.
+
+##### Challenges
+
+Some of the identfied challenges are:
```
Comment (severindellsperger) [inline README.md]:
```suggestion
Some of the identified challenges are:
```

File: README.md [hunk 7]
```diff
@@ -1,57 +1,348 @@
...
+1. Mapping of energy metrics to comparable values
+1. Flexibility in regards to environmental factors 
+1. Knowledge about the path a specific path metric belongs to
+
+The challenges where adressed as follows:
+
+1. To ensure data can be processed at line rate the efficiency data is collected outside of the data plane. It is supposed to be exposed to the data plane via a lookup table comparable to the tables of the forwarding information base (FIB). Additionally the aggregation which happens inside the data plane avoids using complex arithmetical operations.
+1. The mitigation of the increased packet size due to INT goes hand in hand with the challenge/requirement to aggregate the data in transit which reduces the volume of network telemetry data present in packet headers.
+1. To aggregate the efficiency data in transit the [IOAM Aggregation Trace Option](https://datatracker.ietf.org/doc/html/draft-cxx-ippm-ioamaggr-02) is used.
+1. As shown in figure (todo reference) one or more energy metrics are mapped to a hop metric. With a consistent energy metric to hop metric configuration on all nodes the hop metrics of the same type will be comparable.
```
Comment (severindellsperger) [inline README.md]: Add reference

File: README.md [hunk 8]
```diff
@@ -1,57 +1,348 @@
...
+1. The energy metric mapping as shown in figure (todo reference) can be setup as required but needs to be consistent throughout the same IOAM domain. In case energy metrics shall also be exchanged between autonomous systems (AS) with e.g. BGP in future some common configurations need to be standardized.
```
Comment (severindellsperger) [inline README.md]: Add reference

File: README.md [hunk 9]
```diff
@@ -1,57 +1,348 @@
...
+As already stated earlier the BMv2 software switches are used in the PoC network environment.
+By default the BMv2 switches are not capable of exporting data via IPFIX.
+A fork with the corresponding IPFIX implementation is available [here](https://github.com/ramobis/behavioral-model).
+The implementation can be found in the `externs` directory.
+
+There are two different export mechanisms which fulfill different needs.
+
+##### Aggregated Export
+
+The aggregated export is used to make **flow statistics** available to a network operator emphasizing the capability of the generation of a heat map containing endpoint to endpoint network efficiency information.
```
Comment (severindellsperger) [inline README.md]: Maybe link a picture to illustrate the heatmap.

File: README.md [hunk 10]
```diff
@@ -1,57 +1,348 @@
...
+
+As part of the aggregated export data is grouped by flow and IOAM aggregator.
+The network telemetry data of all packets within the same group is then aggregated using the IOAM aggregator specified and exported within a single IPFIX message.
+
+##### Raw Export
+
+The raw export is used to make **path statistics** available to a network operator emphasizing the identification possibilities of inefficient paths and nodes within a network.
```
Comment (severindellsperger) [inline README.md]: Here,, it's also possible to refer to a picture.

File: README.md [hunk 11]
```diff
@@ -1,57 +1,348 @@
...
+
+With this export mechanism the complete IPv6 header including all extension headers is exported.
+That means that not only the efficiency metric is going to be available but also the corresponding path information the metric belongs to.
+
+The raw export is implemented in such a way that it always exports the first packet of a flow and then every nth packet based on the configured sampling rate per flow.
+
+#### Dashboard
+
+The Grafana dashboards visualize the time series data present in InfluxDB.
```
Comment (severindellsperger) [inline README.md]: You could add an example picture, which can be referenced in the introduction.

File: README.md [hunk 12]
```diff
@@ -1,57 +1,348 @@
...
+The purpose of this dashboards are that a human can interpret with little mental effort how the network as a whole, indivdual paths, indivual nodes and individual flows perform in the context of the sustainability metrics.
+It is also helpful to see how the metrics change over time.
+**During the hackathon these dashaboards will be very helpful to see if the optimizer application actually increases the efficiency of the network.**
+
+##### Simulation Network Statistics
```
Comment (severindellsperger) [inline README.md:190]: Reference pictures so the user can look at the dashboards.

File: README.md [hunk 13]
```diff
@@ -1,57 +1,348 @@
...
 
-### Information to share with other participants
-To ensure effective collaboration, we recommend reviewing the paper referenced in the project proposal for a deeper understanding of our approach.
-It would also be helpful to familiarize yourself with the IOAM protocol, particularly the IOAM Aggregation Trace Option, which is a key element in collecting energy efficiency data in the approach proposed.
-You can find the details of this extension in the [Aggregation Trace Option for In-situ Operations, Administration, and Maintenance (IOAM)](https://datatracker.ietf.org/doc/draft-cxx-ippm-ioamaggr/02/) specification.
+This includes the following work items:
+
+- Analysis of given raw data in InfluxDB  
+- **Development of green-path-optimizer application which suggests path updates to improve efficiency**
+- **Integration of the configuration-update-utility to acutally deploy the improvements in the PoC network environment**
```
Comment (severindellsperger) [inline README.md]:
```suggestion
- **Integration of the configuration-update-utility to actually deploy the improvements in the PoC network environment**
```

File: README.md [hunk 14]
```diff
@@ -1,57 +1,348 @@
...
+
+- `config:` Executes the `generate-config` and then the `update-config` target.
+- `generate-config:` Regenerates all configuration files based on the resource file specified in the Makefile.
+This task is automatically executed as part of the `run` target.
+- `update-config:` Pushes the generated configurations to the BMv2 software switches.
+Make sure to select the correct inventory in the `config.yaml` file inside the config updater.
+
+### Stopping Environment
+
+To stop the environment there are dedicated stop targets.
```
Comment (severindellsperger) [inline README.md]: Mention `make`

File: README.md [hunk 15]
```diff
@@ -1,57 +1,348 @@
...
+
+- `stop:` Stops the network and the monitoring environment
+- `stop-network:` Stops the network environment
+- `stop-monitoring:` Stops the monitoring environment
+
+### Cleaning Environment
+
+To clean the environment there are dedicated clean targets.
```
Comment (severindellsperger) [inline README.md]: Mention `make`

File: config/updater/README.md
Comment (severindellsperger) [inline config/updater/README.md:1]: Do you want to use the same readme as in the bachelor thesis?

File: monitoring/README.md
Comment (severindellsperger) [inline monitoring/README.md:1]: Do you want to use it like in the bachelor's thesis? Is this readme referenced in the main readme?

**Review by severindellsperger — COMMENTED** [review]
Overall, it looks good to me. Please address my comments, especially since some typos need to be corrected.

File: README.md (hunk 5 above)
Comment (ramobis) [reply README.md:51]: Good idea!
I will do that.

File: config/updater/README.md
Comment (ramobis) [reply config/updater/README.md:1]: Ups forgot to delete those.
Thank you!

File: monitoring/README.md
Comment (ramobis) [reply monitoring/README.md:1]: removed

File: README.md (hunk 5 above)
Comment (ramobis) [reply README.md:51]: done

**Review by severindellsperger — APPROVED** [review]
LGTM

**Pull request merged**